### PR TITLE
[BugFix] Hide public repos owned by private orgs

### LINKED ISSUE #9555: [bug] global explore show repos of hidden org
Bildschirmfoto zu 2019-12-31 00-39-16

if you like to look at it - you get a 404 wich shows that the repo view handle rights in a correct way

### PR DESCRIPTION
**Pull request by 6543**
fix #9555

### CODE REVIEW

File: models/repo_list.go [hunk 1]
```diff
@@ -228,7 +228,7 @@ func SearchRepository(opts *SearchRepoOptions) (RepositoryList, int64, error) {
 		}
 
 		if opts.AllPublic {
-			accessCond = accessCond.Or(builder.Eq{"is_private": false})
+			accessCond = accessCond.Or(builder.Eq{"is_private": false}).And(builder.In("owner_id", builder.Select("id").From("user").Where(builder.Lte{"visibility": 1})))
```
Comment (davidsvantesson) [inline models/repo_list.go]: Does this work for user repositories?

Comment (6543) [reply models/repo_list.go]: sure, default value is 0 and user dont use this flag so user are always public

Comment (6543) [reply models/repo_list.go]: And if we implement visabilety for users, it will handle out of the box ;)

**Review by davidsvantesson — APPROVED** [review]

**Comment by 6543** [comment]
since i use xorm builder and dont use any sql statement it should be a **xorm** error ?
@lunny 

`SearchRepository: Count: mssql: Incorrect syntax near the keyword 'user'.`
and
`SearchRepository: Count: pq: column "visibility" does not exist`

File: models/repo_list.go (hunk 1 above)
Comment (6543) [inline models/repo_list.go]:
```suggestion
			accessCond = accessCond.Or(builder.Eq{"is_private": false}).And(builder.In("owner_id", builder.Select("user.id").From("user").Where(builder.Lte{"user.visibility": 1})))
```

**Comment by 6543** [comment]
@lunny found the bug `user` a reserved word like index so I had to add '`'

**Comment by 6543** [comment]
@lunny b689efbd945c157f783b42703a784b00a4116a9e did it :)

can you :eyes:  at this bugfix?

File: models/repo_list.go [hunk 2]
```diff
@@ -228,7 +228,7 @@ func SearchRepository(opts *SearchRepoOptions) (RepositoryList, int64, error) {
 		}
 
 		if opts.AllPublic {
-			accessCond = accessCond.Or(builder.Eq{"is_private": false})
+			accessCond = accessCond.Or(builder.Eq{"is_private": false}).And(builder.In("owner_id", builder.Select("`user`.id").From("`user`").Where(builder.Lte{"`user`.visibility": 1})))
```
Comment (6543) [inline models/repo_list.go]:
```suggestion
			accessCond = accessCond.Or(builder.Eq{"is_private": false}).And(builder.In("owner_id", builder.Select("`user`.id").From("`user`").Where(builder.Lte{"`user`.visibility": structs.VisibleTypePrivate})))
```

File: models/repo_list.go [hunk 3]
```diff
@@ -228,7 +228,7 @@ func SearchRepository(opts *SearchRepoOptions) (RepositoryList, int64, error) {
 		}
 
 		if opts.AllPublic {
-			accessCond = accessCond.Or(builder.Eq{"is_private": false})
+			accessCond = accessCond.Or(builder.Eq{"is_private": false}).And(builder.In("owner_id", builder.Select("`user`.id").From("`user`").Where(builder.Lte{"`user`.visibility": structs.VisibleTypePrivate})))
```
Comment (6543) [inline models/repo_list.go]:
```suggestion
			accessCond = accessCond.Or(builder.Eq{"is_private": false}).And(builder.In("owner_id", builder.Select("`user`.id").From("`user`").Where(builder.Lte{"`user`.visibility": structs.VisibleTypeLimited})))
```

**Comment by davidsvantesson** [comment]
The issue title has an extra d in "Hide", maybe you can fix that.

**Comment by 6543** [comment]
efa0f5461bf44e07e777ddff1a3978d8355cc169 :sweat_smile: if you convert `1` to enum you should use the right value :D

@davidsvantesson I changed the query a bit - hope It is still ok ...?

File: models/repo_list.go [hunk 4]
```diff
@@ -228,7 +228,7 @@ func SearchRepository(opts *SearchRepoOptions) (RepositoryList, int64, error) {
 		}
 
 		if opts.AllPublic {
-			accessCond = accessCond.Or(builder.Eq{"is_private": false})
+			accessCond = accessCond.Or(builder.Eq{"is_private": false}).And(builder.In("owner_id", builder.Select("`user`.id").From("`user`").Where(builder.Lte{"`user`.visibility": structs.VisibleTypeLimited})))
```
Comment (davidsvantesson) [inline models/repo_list.go]: Do we know that someone is logged in here, otherwise Limited shouldn't be shown?

File: models/repo_list_test.go [hunk 5]
```diff
@@ -178,7 +178,7 @@ func TestSearchRepository(t *testing.T) {
 			count: 25},
 		{name: "AllPublic/PublicAndPrivateRepositoriesOfUserIncludingCollaborative",
 			opts:  &SearchRepoOptions{Page: 1, PageSize: 10, OwnerID: 15, Private: true, AllPublic: true, Template: util.OptionalBoolFalse},
-			count: 31},
+			count: 30},
```
Comment (bagasme) [inline models/repo_list_test.go:181]: `count` decrease?

Comment (6543) [reply models/repo_list_test.go:181]: Of course now we hide a public repo of a private org

File: models/repo_list.go (hunk 4 above)
Comment (6543) [reply models/repo_list.go]: let me look otherwhise we need more refactor :/

**Comment by 6543** [comment]
@davidsvantesson to "Do we know that someone is logged in here, otherwise Limited shouldn't be shown?"

I did a refactor of the code to -> ef3e985676e856fde76c669bb9d11a8a69e7d41c

**Review by davidsvantesson — APPROVED** [review]

File: routers/home.go [hunk 6]
```diff
@@ -142,6 +142,7 @@ func RenderRepoSearch(ctx *context.Context, opts *RepoSearchOptions) {
 		Keyword:            keyword,
 		OwnerID:            opts.OwnerID,
 		AllPublic:          true,
+		AllLimited:         true,
```
Comment (lafriks) [inline routers/home.go:145]: Shouldn't this be true only if user is authorized?

Comment (6543) [reply routers/home.go]: yes - and in this functionroutine it is :) @lafriks

**Comment by zeripath** [comment]
Hmm... I'm concerned that this might not be the correct place. 

https://github.com/go-gitea/gitea/blob/a5a79f702d79b4288d6b214fc9f83ed725edf385/models/repo_list.go#L181-L196

is supposed to limit the repositories to things that can be seen already... So if explore is displaying an inaccessible repo then that's where the problem lies.

**Comment by zeripath** [comment]
OK so `/explore/repos` points to:

https://github.com/go-gitea/gitea/blob/7e4f490482014fa52d15902b5aa20952461c64f4/routers/home.go#L165-L183

Which means that in the non-logged in case you get:

```go
RenderRepoSearch(ctx, &RepoSearchOptions{ 
 		PageSize: setting.UI.ExplorePagingNum, 
 		OwnerID:  0, 
 		Private:  false, 
 		TplName:  tplExploreRepos, 
 	})
```

which should map to the false branch above which should be correct and the opts.PublicAll doesn't apply.

**Comment by 6543** [comment]
OwnerID:  0  && Private:  false

then only this code create the condition:

https://github.com/go-gitea/gitea/blob/master/models/repo_list.go#L186-L195

**Comment by 6543** [comment]
@zeripath there was no bug with **loged out** users but if you are **loged In** you could saw the public repos of private organisations

**Comment by zeripath** [comment]
Then there's a problem with https://github.com/go-gitea/gitea/blob/a5a79f702d79b4288d6b214fc9f83ed725edf385/models/repo_list.go#L181-L186

e.g. with:

https://github.com/go-gitea/gitea/blob/a5a79f702d79b4288d6b214fc9f83ed725edf385/models/repo_list.go#L311-L338

**Comment by zeripath** [comment]
You should only get repositories that you able to see from that condition.

**Comment by zeripath** [comment]
ah I am beginning to re-understand this bloody weird code.

**Review by zeripath — APPROVED** [review]

**Comment by zeripath** [comment]
make lg-tm work.

**Comment by zeripath** [comment]
I think that function needs refactoring to always restrict to the accessible repos for the current user.

**Comment by 6543** [comment]
@zeripath yes this need a refactor ...
schould i backport?

**Pull request merged**